src/main/java/org/opentripplanner/routing/trippattern/TripTimes.java

```diff
@@ -474,6 +478,66 @@ public boolean tripAcceptable(final State state0, final int stopIndex) {
         return true;
     }
 
+    /**
+     * Returns true if the current state chain contains a banned trip, or banned trip/stop index or banned trip sequence
+     */
+    public boolean tripOrTripSequenceIsBanned(State state0, final int stopIndex) {
+        final RoutingRequest options = state0.getOptions();
+
+        // check if an exact trip ID is banned
+        final BannedStopSet banned = options.bannedTrips.get(trip.getId());
+        if (banned != null && banned.contains(stopIndex)) {
+            return true;
+        }
+
+        // check if a trip ID sequence is banned, but first return false if there are not banned sequences
+        if (options.bannedTripSequences.size() == 0) {
+            return false;
+        }
+
+        // compile list of trips seen so far in states (use ArrayList constructor to ensure a mutable list is returned)
+        List<FeedScopedId> tripsInState = new ArrayList<>(Arrays.asList(trip.getId()));
+        State tripFindingState = state0;
+        Trip lastTrip = trip;
+        while (tripFindingState != null) {
+            if (tripFindingState.getBackEdge() != null) {
+                Trip trip = tripFindingState.getBackTrip();
+                if (trip != null && trip != lastTrip) {
+                    tripsInState.add(trip.getId());
+                    lastTrip = trip;
+                }
+            }
+            tripFindingState = tripFindingState.getBackState();
+        }
+
+        // reverse trips if finding for depart-at search to make sure the list is in chronological order
+        if (!options.arriveBy) {
+            Collections.reverse(tripsInState);
+        }
+
+        // Check for the trip state sequence in all banned trip sequences. This will return true even for sequences that
+        // have additional trips included between a banned sequence. (ex: if trip A > trip B is banned, trip A >
+        // Trip C > Trip B will be flagged as a banned sequence).
+        for (List<FeedScopedId> bannedTripSequence : options.bannedTripSequences) {
+            int tripsInStateIdx = 0;
+            int bannedSequenceIdx = 0;
```

Comment:
To reduce wordiness of this whole block I might rename these to

```java
int seenIdx = 0;
int bannedIdx = 0;
```

Reply:
This would confuse the heck out of me. What is `seen`? The `tripMatches` isn't appropriate because this variable is used specifically to track indexes in a list and retrieve items from the list according to the current index. The check at the end is confirming the number of trip matches, but by comparison of the current index to the size.

Reply:
I ended up using Binh's refactor suggestion (https://github.com/ibi-group/OpenTripPlanner/pull/52#discussion_r597755015) in 6adec27faa2e744f94cd91e4ae4d337a5c25c7c7.